### PR TITLE
Adapting to the BatchNorm structure to support 3D data

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/3597
Adapting to the BatchNorm structure to support 3D data

### CODE REVIEW

File: python/paddle/trainer_config_helpers/tests/layers_test.py [hunk 1]
```diff
@@ -16,4 +16,4 @@
 
 if __name__ == '__main__':
     parse_config_and_serialize(
-        'trainer_config_helpers/tests/layers_test_config.py', '')
+        'trainer_config_helpers/tests/configs/test_BatchNorm3D.py', '')
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/tests/layers_test.py]: Do not change this line. Add `test_BatchNorm3D.py` in the file `python/paddle/trainer_config_helpers/tests/configs/file_list.sh`. 

And please merge 3DConv/3DPooling first, then update this PR.

Comment (chengduoZH) [reply python/paddle/trainer_config_helpers/tests/layers_test.py]: Done

File: proto/ModelConfig.proto [hunk 2]
```diff
@@ -515,7 +516,8 @@ message LayerConfig {
   // for HuberRegressionLoss
   optional double delta = 57 [ default = 1.0 ];
 
-  optional uint64 depth = 58 [ default = 1 ];
+  // for 3D data
+  optional double depth = 58 [ default = 1 ];
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply proto/ModelConfig.proto]: Sorry, It has been fixed

File: python/paddle/trainer/config_parser.py [hunk 3]
```diff
@@ -2688,9 +2735,20 @@ def __init__(self, name, inputs, bias=True, **xargs):
...
-            self.set_layer_size(input_layer.size)
+
+        if len(self.inputs) > 1:
+            for input_index in xrange(len(self.inputs)):
+                assert self.get_input_layer(0).height == self.get_input_layer(
+                    input_index).height
+                assert self.get_input_layer(0).width == self.get_input_layer(
+                    input_index).width
+                assert self.get_input_layer(0).depth == self.get_input_layer(
+                    input_index).depth
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: 严格来说，对于Addto，我觉得只需要检测input_layer.size相等即可。

File: python/paddle/trainer/config_parser.py [hunk 4]
```diff
@@ -3370,11 +3428,20 @@ def __init__(self, name, inputs, bias=False, **xargs):
...
+            assert self.get_input_layer(0).depth == self.get_input_layer(
+                input_index).depth
             input_layer = self.get_input_layer(input_index)
             input = self.inputs[input_index]
             if self.config.size == 0:
                 size += input_layer.size
 
+        self.set_layer_height_width(self.get_input_layer(0).height, \
+                                    self.get_input_layer(0).width)
+        self.set_layer_depth(self.get_input_layer(0).depth)
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: 实际上不只Python这边需要对AddtoLayer  ConcatLayer设置height width depth。C++的实现也需要set一下，这样才能支持变长输入。

**Pull request merged**